contracts/OlasMech.sol

```diff
@@ -90,118 +97,88 @@ abstract contract OlasMech is Mech, IErrorsMech, ImmutableStorage {
-        // Check if the request Id is invalid (non existent or delivered): previous and next request Ids are zero,
-        // and the zero's element previous request Id is not equal to the provided request Id
-        if (requestIds[0] == 0 && requestIds[1] == 0 && mapRequestIds[0][0] != requestId) {
-            revert RequestIdNotFound(requestId);
-        }
-
-        // Re-link previous and next elements between themselves
-        mapRequestIds[requestIds[0]][1] = requestIds[1];
-        mapRequestIds[requestIds[1]][0] = requestIds[0];
-
-        // Delete the delivered element from the map
-        delete mapRequestIds[requestId];
+        numTotalRequests += numRequests;
     }
 
-    /// @dev Delivers a request.
+    /// @dev Prepares delivery of requests.
     /// @notice This function ultimately calls mech marketplace contract to finalize the delivery.
-    /// @param requestId Request id.
-    /// @param data Self-descriptive opaque data-blob.
-    function _deliver(uint256 requestId, bytes memory data) internal virtual returns (bytes memory requestData) {
-        // Get an requester to deliver request to
-        address requester = mapRequestAddresses[requestId];
-
-        // Get the mech delivery info from the mech marketplace
-        IMechMarketplace.MechDelivery memory mechDelivery =
-            IMechMarketplace(mechMarketplace).mapRequestIdDeliveries(requestId);
-
-        // Instantly return with empty data if the request has been delivered
-        // This allows not to fail batch requests transactions
-        if (mechDelivery.deliveryMech != address(0)) {
-            return "";
-        }
-
-        // The requester is zero if the delivery mech is different from a priority mech, or if request does not exist
-        if (requester == address(0)) {
-            requester = mechDelivery.requester;
-            // Check if request exists in the mech marketplace
-            if (requester == address(0)) {
-                revert RequestIdNotFound(requestId);
+    /// @param requestIds Set of request Ids.
+    /// @param datas Corresponding set of self-descriptive opaque delivery data-blobs.
+    /// @return deliveryDatas Corresponding set of processed delivery datas.
+    function _prepareDeliveries(
+        uint256[] memory requestIds,
+        bytes[] memory datas
+    ) internal virtual returns (bytes[] memory deliveryDatas) {
+        uint256 numRequests = requestIds.length;
+        deliveryDatas = new bytes[](numRequests);
+
+        uint256 numSelfRequests;
+        // Traverse requests
+        for (uint256 i = 0; i < numRequests; ++i) {
+            uint256 requestId = requestIds[i];
+            // Perform a pre-delivery of the data if it needs additional parsing
+            deliveryDatas[i] = _preDeliver(requestId, datas[i]);
+
+            // Clean request info
+            // Get request Id from the request Ids map
+            uint256[2] memory requestIdLinks = mapRequestIds[requestId];
```

Comment:
Check if request was assigned to this mech, and if it was - reduce the undelivered count of this mech.